src/transforms/remap.rs

```diff
@@ -436,6 +467,107 @@ where
             }
         }
     }
+
+    fn transform_all(&mut self, events: EventArray, output: &mut TransformOutputsBuf) {
+        let events_len = events.len();
+        let output_capacity = output.capacity();
+        if events_len > output_capacity {
+            output.reserve(events_len - output_capacity);
+        }
+
+        // If a program can fail or abort at runtime and we know that we will still need to forward
+        // the event in that case (either to the main output or `dropped`, depending on the
+        // config), we need to clone the original event and keep it around, to allow us to discard
+        // any mutations made to the event while the VRL program runs, before it failed or aborted.
+        //
+        // The `drop_on_{error, abort}` transform config allows operators to remove events from the
+        // main output if they're failed or aborted, in which case we can skip the cloning, since
+        // any mutations made by VRL will be ignored regardless. If they have configured
+        // `reroute_dropped`, however, we still need to do the clone to ensure that we can forward
+        // the event to the `dropped` output.
+        let forward_on_error = !self.drop_on_error || self.reroute_dropped;
+        let forward_on_abort = !self.drop_on_abort || self.reroute_dropped;
+        let mut original_events = if (self.program.info().fallible && forward_on_error)
+            || (self.program.info().abortable && forward_on_abort)
+        {
+            Some(events.clone())
+        } else {
+            None
+        };
+
+        let program_info = self.program.info();
+        let mut targets = events
+            .into_events()
+            .map(|event| VrlTarget::new(event, program_info))
+            .collect::<Vec<_>>();
+        let results = self.run_vrl_batch(&mut targets);
+
+        for (i, (result, target)) in results.into_iter().zip(targets.into_iter()).enumerate() {
```

Comment:
Is there any way to `zip` this with `original_events` to avoid the `take` and indexing games below?